pkg/services/sqlstore/migrations/accesscontrol/dashboard_permissions.go

```diff
@@ -661,6 +661,174 @@ func (m *managedFolderLibraryPanelActionsMigrator) Exec(sess *xorm.Session, mg *
+		return err
+	}
+
+	var ids []any
+	if err := sess.SQL("SELECT id FROM role WHERE name LIKE 'managed:%'").Find(&ids); err != nil {
+		return err
+	}
+
+	if len(ids) == 0 {
+		return nil
+	}
+
+	var permissions []ac.Permission
+	roleQueryBatchSize := 100
+	batch(len(ids), roleQueryBatchSize, func(start, end int) error {
+		var batchPermissions []ac.Permission
+		if err := sess.SQL("SELECT role_id, action, scope FROM permission WHERE role_id IN(?"+strings.Repeat(" ,?", len(ids[start:end])-1)+") AND (scope LIKE 'folders:%' or scope LIKE 'dashboards:%')", ids[start:end]...).Find(&batchPermissions); err != nil {
+			return err
+		}
+		permissions = append(permissions, batchPermissions...)
+		return nil
+	})
+
+	mapped := make(map[int64]map[string]map[string]bool, len(ids)-1)
+	for _, p := range permissions {
+		if mapped[p.RoleID] == nil {
+			mapped[p.RoleID] = make(map[string]map[string]bool)
+		}
+		if mapped[p.RoleID][p.Scope] == nil {
+			mapped[p.RoleID][p.Scope] = make(map[string]bool)
+		}
+		mapped[p.RoleID][p.Scope][p.Action] = true
+	}
+
+	var toAdd []ac.Permission
+	now := time.Now()
+
+	for roleId, mappedPermissions := range mapped {
+		for scope, roleActions := range mappedPermissions {
+			if roleActions[dashboards.ActionDashboardsRead] {
+				if !roleActions[ac.ActionAnnotationsRead] {
+					toAdd = append(toAdd, ac.Permission{
+						RoleID:  roleId,
+						Updated: now,
+						Created: now,
+						Scope:   scope,
+						Action:  ac.ActionAnnotationsRead,
+					})
+				}
+			}
+
+			if roleActions[dashboards.ActionDashboardsWrite] {
+				if !roleActions[ac.ActionAnnotationsCreate] {
+					toAdd = append(toAdd, ac.Permission{
+						RoleID:  roleId,
+						Updated: now,
+						Created: now,
+						Scope:   scope,
+						Action:  ac.ActionAnnotationsCreate,
+					})
+				}
+				if !roleActions[ac.ActionAnnotationsDelete] {
+					toAdd = append(toAdd, ac.Permission{
+						RoleID:  roleId,
+						Updated: now,
+						Created: now,
+						Scope:   scope,
+						Action:  ac.ActionAnnotationsDelete,
+					})
+				}
+				if !roleActions[ac.ActionAnnotationsWrite] {
+					toAdd = append(toAdd, ac.Permission{
+						RoleID:  roleId,
+						Updated: now,
+						Created: now,
+						Scope:   scope,
+						Action:  ac.ActionAnnotationsWrite,
+					})
+				}
+			}
+		}
+	}
+
+	if len(toAdd) == 0 {
+		return nil
+	}
+
+	return batch(len(toAdd), batchSize, func(start, end int) error {
+		_, err := sess.InsertMulti(toAdd[start:end])
+		return err
+	})
+}
+
+func (m *managedDashboardAnnotationActionsMigrator) hasDefaultAnnotationPermissions(sess *xorm.Session, mg *migrator.Migrator) (bool, error) {
+	type basicRolePermission struct {
+		Uid    string
+		Action string
+		Scope  string
+	}
+
+	var basicRolePermissions []basicRolePermission
+	basicRoleUIDs := []any{"basic_viewer", "basic_editor", "basic_admin"}
+	query := `SELECT r.uid, p.action, p.scope FROM role r
```

Comment:
Yeees, I think this would work. The logged into would be potentially less detailed, but I think this would work in general. It should be `4,4,4` though, right? I'll refactor the check now.

Reply:
Oh no, you're right about the numbers. I think I'll need to add a `distinct` to make sure that all the different expected actions are present.

Reply:
I thought about this some more, and I worry that we might miss some really weird edge cases if we only look at the numbers. For instance, if someone has added `annotations:read` to a basic role several times (eg, by manually adding `annotations:read` to it with different wildcard scopes), but has also removed, say, `annotations:create`. Because we only see the numbers, we don't know what actions lie behind the scopes. So I'm a bit scared we might have some false positives, even if it's unlikely. There's probably some clever way of grouping to account for this, but my brain is too tired for it now.

Reply:
Sounds good to me. :)